src/tree.rs

```diff
@@ -166,6 +168,55 @@ impl<T: Default> Tree<T> {
     }
 }
 
+impl Tree<Item> {
+    /// Truncates the preceding siblings to the given end position,
+    /// and returns the new current node.
+    pub(crate) fn truncate_siblings(&mut self, end_byte_ix: usize) -> Option<TreeIndex> {
```

Comment:
This is the only part of the PR I don't fully understand. Why do we need this? I understand we need to truncate the header texts, but it's all one node right? It doesn't get split into multiple nodes until the second parsing phase. Why can't we just adjust the header end?

Reply:
It is because:

1. Truncating prefix of the header before parsing seems easy, but truncating suffix would be non-trivial.
2. Parsing by `FirstPass::parse_line` seems to include some preparation for the second pass.

Consider the input is `# **Foo {#bar** }`.
This is ATX heading so `FirstPass::parse_atx_heading` is called.
At line 1012 on firstpass.rs (master branch), `text[ix..]` is `**Foo {#bar** }`, and the parser tries to parse and get a line.

https://github.com/raphlinus/pulldown-cmark/blob/d412d7cb652707f8c980c59dca836a6ad4e49f2f/src/firstpass.rs#L1012

This is, (1) truncating prefix is easy but truncating suffix won't.

Then, [`FirstPass::parse_line`](https://github.com/raphlinus/pulldown-cmark/blob/d412d7cb652707f8c980c59dca836a6ad4e49f2f/src/firstpass.rs#L408} actually does more job than getting a line.
It creates a tree for the input `**Foo {#bar** }...` as below:

* MaybeEmphasis (1st, can open)
* MaybeEmphasis (2nd, can open)
* Text (`Foo {#bar`)
* MaybeEmphasis (1st, can close)
* MaybeEmphasis (2nd, can close)
* Text (` }`)

Second pass utilizes this `MaybeEmphasis`'s to create real tree structure.
This is, (2) parsing by `FirstPass::parse_line` seems to include some preparation for the second pass.

* * *

In my current implementation, I use `parse_line` same way as before.
And after that parsing, modify the (relatively) flat tree into truncated form, such as below:

* MaybeEmphasis (1st, can open)
* MaybeEmphasis (2nd, can open)
* Text (`Foo`)

This is what I wanted to say by https://github.com/raphlinus/pulldown-cmark/pull/522#issuecomment-973204747:
> For example, this new method is used when turning `__`/`foo {.bar`/`__`/`}` into `__`/`foo`, since `{.bar__}` suffix should be removed as an attribute block.

* * *

Ideally, the suffix truncation should be done before `parse_line`, but it seems `parse_line` only accepts half-open byte slice.
Making it accept closed range of the bytes might be large diff, and I tried to avoid making such big change that would affect every part of the parsing.

Should I modify `parse_line` to accept not only `start: usize` but also `end: Option<usize>`?

Reply:
Thank you for the detailed explanation, I had not considered that `parse_line` already adds these tree elements! It makes total sense why you do these things now.

However, I do think that it would be cleaner to scan for the attribute block first (if the option is set!) and only after that calling the `parse_line` function on the prefix. For two reasons: we wouldn't waste effort parsing the header as if it were normal markdown, and truncating the header text would become trivial.

We would indeed have to adjust `parse_line` it seems, and adding an optional end the way you suggest looks like a good way of doing it! 👍🏽

Reply:
I tried rewriting `parse_atx_heading()` without `truncate_siblings()` (by trimming the attribute block before parsing the line) and succeeded.

However, it seems hard to do the same for `parse_setext_heading()` since it transforms an already-parsed paragraph into a heading.
At the time lines are being parsed, the parser cannot know whether they are setext heading content.
So it will be necessary to modify already-parsed tree in that case and I'll continue to use `truncate_siblings()` for setext heading.